File: backend/src/lib/normalize.ts
import type { DecisionEvent, GraphEdge } from '../schemas/decision.schema';
import {
  agentNodeId,
  graphIdFor,
  parseGraphId,
  primaryUserNodeId,
  resolveAgentContext,
  userSlugFromNodeId,
} from './graph-topology';

const LEGACY_AGENT_ID = 'publisher_agent';
const AGENT_DISPLAY_ID = 'Publisher Agent';

const LABEL_MAP: Record<string, string> = {
  user: 'User',
  protocol: 'Protocol',
  market: 'Market',
  correlated_market: 'CorrelatedMarket',
  trade: 'Trade',
  outcome: 'Outcome',
  feedback: 'Feedback',
  agent: 'Agent',
};

/** Legacy Sell decision id -> original Buy trade id for the same position. */
const LEGACY_CLOSE_TRADE_OF: Record<string, string> = {
  TRD_003: 'TRD_001',
  TRD_005: 'TRD_004',
  TRD_007: 'TRD_006',
  TRD_010: 'TRD_008',
  TRD_012: 'TRD_009',
  TRD_014: 'TRD_011',
  TRD_017: 'TRD_013',
};

function slug(value: string): string {
  return value.replace(/[^a-zA-Z0-9]+/g, '_').replace(/^_|_$/g, '').toLowerCase();
}

function actionToRel(action: string): string {
  return action.trim().toUpperCase().replace(/\s+/g, '_');
}

function inferDecisionId(edges: GraphEdge[], updatedAt: string): string {
  for (const edge of edges) {
    if (edge.Action && edge.target) return `dec-${slug(edge.target)}`;
  }
  return `dec-${updatedAt.replace(/[:.-]/g, '')}`;
}

export function graphName(graphId: string): string {
  return graphId.replace(/[^a-zA-Z0-9_]/g, '_');
}

function remapLegacyAgent(payload: DecisionEvent, agentId: string): void {
  const nodeIds = new Set(payload.nodes.map((n) => n.node_id));
  if (!nodeIds.has(LEGACY_AGENT_ID) && !nodeIds.has(agentId)) {
    const refsLegacy = payload.edges.some(
      (e) =>
        e.source === LEGACY_AGENT_ID ||
        e.target === LEGACY_AGENT_ID ||
        e.source === AGENT_DISPLAY_ID ||
        e.target === AGENT_DISPLAY_ID,
    );
    if (refsLegacy) {
      payload.nodes.push({
        node_id: agentId,
        node_type: 'agent',
        properties: { display_name: 'Publisher Agent', role: 'publisher' },
        label: 'Agent',
      });
    }
  }

  for (const node of payload.nodes) {
    if (node.node_id === LEGACY_AGENT_ID) node.node_id = agentId;
  }

  for (const edge of payload.edges) {
    if (edge.source === LEGACY_AGENT_ID || edge.source === AGENT_DISPLAY_ID) edge.source = agentId;
    if (edge.target === LEGACY_AGENT_ID || edge.target === AGENT_DISPLAY_ID) edge.target = agentId;
    if (edge.targets) {
      edge.targets = edge.targets.map((t) =>
        t === LEGACY_AGENT_ID || t === AGENT_DISPLAY_ID ? agentId : t,
      );
    }
  }
}

function remapNodeId(payload: DecisionEvent, from: string, to: string): void {
  const node = payload.nodes.find((n) => n.node_id === from);
  if (!node || from === to) return;
  node.node_id = to;
  for (const edge of payload.edges) {
    if (edge.source === from) edge.source = to;
    if (edge.target === from) edge.target = to;
    if (edge.targets) {
      edge.targets = edge.targets.map((t) => (t === from ? to : t));
    }
  }
}

function collapseAgentIntoUser(payload: DecisionEvent, userNodeId: string, agentId: string): void {
  if (userNodeId === agentId) return;

  for (const edge of payload.edges) {
    if (edge.source === agentId) edge.source = userNodeId;
    if (edge.target === agentId) edge.target = userNodeId;
    if (edge.targets) {
      edge.targets = edge.targets.map((t) => (t === agentId ? userNodeId : t));
    }
  }

  payload.edges = payload.edges.filter((e) => e.relationship_type !== 'HAS_AGENT');
  payload.nodes = payload.nodes.filter((n) => n.node_id !== agentId && n.node_type !== 'agent');
}

/** Stable per-trade leaf id; safe if normalize runs more than once (seed + worker). */
export function scopedLeafId(leafId: string, tradeId: string): string {
  const suffix = `__${tradeId}`;
  let base = leafId;
  while (base.endsWith(suffix)) base = base.slice(0, -suffix.length);
  return `${base}${suffix}`;
}

function normalizeLegacyBuySellLifecycle(payload: DecisionEvent): void {
  for (const edge of payload.edges) {
    if (typeof edge.Action !== 'string' || !edge.target) continue;
    const match = edge.Action.match(/^(Buy|Sell)\s+(YES|NO)$/i);
    if (!match) continue;

    const verb = match[1].toLowerCase();
    const side = match[2].toUpperCase();
    const originalTradeId = edge.target;
    const isClose = verb === 'sell';
    const tradeId = isClose ? (LEGACY_CLOSE_TRADE_OF[originalTradeId] ?? originalTradeId) : originalTradeId;

    if (tradeId !== originalTradeId) {
      remapNodeId(payload, originalTradeId, tradeId);
    }

    edge.Action = `${isClose ? 'Close' : 'Open'} ${side}`;
    edge.relationship_type = actionToRel(edge.Action);
    edge.metadata = {
      ...(edge.metadata ?? {}),
      action: edge.Action,
      lifecycle: isClose ? 'close' : 'open',
      ...(isClose ? { position_trade_id: tradeId } : {}),
    };
    payload.operation = isClose ? 'revise' : 'assert';
    payload.decision_id = `dec-trd_${tradeId.replace('TRD_', '').toLowerCase()}-${
      isClose ? 'close' : 'open'
    }`;
  }
}

/** Open events must not include outcome/feedback nodes or edges (those belong on close only). */
function stripOpenLifecycleOrphans(payload: DecisionEvent): void {
  const isOpen =
    payload.operation === 'assert' &&
    payload.edges.some(
      (e) =>
        e.metadata?.lifecycle === 'open' ||
        (typeof e.Action === 'string' && /^Open\s/i.test(e.Action)),
    );
  if (!isOpen) return;

  const leafIds = new Set(
    payload.nodes
      .filter((n) => n.node_type === 'outcome' || n.node_type === 'feedback')
      .map((n) => n.node_id),
  );

  payload.edges = payload.edges.filter((e) => !(e.target && leafIds.has(e.target)));
  payload.nodes = payload.nodes.filter((n) => {
    if (n.node_type !== 'outcome' && n.node_type !== 'feedback') return true;
    return false;
  });
}

/** Stamp close deltas so ingest can tie back to the same trade node. */
function stampCloseLifecycle(payload: DecisionEvent): void {
  const isClose =
    payload.operation === 'revise' ||
    payload.edges.some(
      (e) =>
        e.metadata?.lifecycle === 'close' ||
        (typeof e.Action === 'string' && /^Close\s/i.test(e.Action)),
    );
  if (!isClose) return;

  const tradeNode = payload.nodes.find((n) => n.node_type === 'trade');
  if (!tradeNode) return;

  for (const e of payload.edges) {
    if (e.Action && e.target === tradeNode.node_id) {
      e.metadata = {
        ...(e.metadata ?? {}),
        lifecycle: 'close',
        position_trade_id: tradeNode.node_id,
      };
    }
  }
}

/** One feedback/outcome node per trade so merged graphs stay U→P→M→T→F/O chains. */
function scopeTradeLeafNodes(payload: DecisionEvent): void {
  const tradeIds = new Set<string>();
  for (const edge of payload.edges) {
    if (edge.Action && edge.target) tradeIds.add(edge.target);
  }

  for (const tradeId of tradeIds) {
    for (const edge of payload.edges) {
      if (edge.source !== tradeId || !edge.target) continue;
      const targetNode = payload.nodes.find((n) => n.node_id === edge.target);
      if (!targetNode) continue;
      if (targetNode.node_type !== 'feedback' && targetNode.node_type !== 'outcome') continue;
      const scopedId = scopedLeafId(edge.target, tradeId);
      if (edge.target !== scopedId) remapNodeId(payload, edge.target, scopedId);
    }
  }
}

/** Markets that open/close a position (source of Action edge → trade). */
export function anchorMarketIds(payload: DecisionEvent): Set<string> {
  const ids = new Set<string>();
  for (const edge of payload.edges) {
    if (!edge.Action || !edge.target) continue;
    const trade = payload.nodes.find((n) => n.node_id === edge.target && n.node_type === 'trade');
    if (trade) ids.add(edge.source);
  }
  return ids;
}

function correlatedTargetIds(payload: DecisionEvent): Set<string> {
  const ids = new Set<string>();
  for (const edge of payload.edges) {
    const rel = (edge.relationship_type ?? String(edge.metadata?.relationship_type ?? ''))
      .toUpperCase()
      .replace(/[^A-Z0-9_]/g, '_');
    if (rel !== 'CORRELATED_MARKET') continue;
    for (const tid of edge.targets ?? []) ids.add(tid);
  }
  return ids;
}

/** Drop correlated peers from nodes[] — they belong in CORRELATED_MARKET targets[] only. */
function stripCorrelatedOnlyMarketNodes(payload: DecisionEvent): void {
  const anchors = anchorMarketIds(payload);
  payload.nodes = payload.nodes.filter((n) => {
    if (n.node_type !== 'market') return true;
    return anchors.has(n.node_id);
  });
}

/**
 * MERGE stub market nodes for correlated targets (not written to decision JSON on disk).
 */
function stampAnchorMarketProperties(payload: DecisionEvent): void {
  const anchors = anchorMarketIds(payload);
  for (const node of payload.nodes) {
    if (node.node_type !== 'market' || !anchors.has(node.node_id)) continue;
    node.properties = { ...(node.properties ?? {}), anchor: true };
  }
}

export function augmentCorrelatedPeerNodes(payload: DecisionEvent): DecisionEvent {
  const existing = new Set(payload.nodes.map((n) => n.node_id));
  const peers = correlatedTargetIds(payload);
  for (const nodeId of peers) {
    if (existing.has(nodeId)) continue;
    payload.nodes.push({
      node_id: nodeId,
      node_type: 'correlated_market',
      properties: { correlated_peer: true },
      label: 'CorrelatedMarket',
    });
    existing.add(nodeId);
  }
  return payload;
}

function ensureHasAgentEdge(payload: DecisionEvent, userNodeId: string, agentId: string, role: string): void {
  const has = payload.edges.some(
    (e) =>
      e.source === userNodeId &&
      e.target === agentId &&
      e.relationship_type === 'HAS_AGENT',
  );
  if (!has) {
    payload.edges.unshift({
      source: userNodeId,
      target: agentId,
      relationship_type: 'HAS_AGENT',
      metadata: { role },
    });
  }
}

export function normalizeDecision(raw: DecisionEvent): DecisionEvent {
  const payload = structuredClone(raw) as DecisionEvent & { edges: GraphEdge[] };
  payload.schema_version = payload.schema_version ?? '1.0';
  payload.operation = payload.operation ?? 'assert';
  normalizeLegacyBuySellLifecycle(payload);

  const ctx = resolveAgentContext(payload);
  if (ctx) {
    if (!parseGraphId(payload.graph_id)) {
      payload.graph_id = graphIdFor(ctx.userSlug, ctx.role);
    }
    remapLegacyAgent(payload, ctx.agentId);
    if (ctx.role === 'seeker') {
      collapseAgentIntoUser(payload, ctx.userNodeId, ctx.agentId);
    } else {
      ensureHasAgentEdge(payload, ctx.userNodeId, ctx.agentId, ctx.role);
    }
  } else {
    const userNodeId = primaryUserNodeId(payload.nodes);
    if (userNodeId) {
      const userSlug = userSlugFromNodeId(userNodeId);
      const agentId = agentNodeId(userSlug, 'publisher');
      remapLegacyAgent(payload, agentId);
      ensureHasAgentEdge(payload, userNodeId, agentId, 'publisher');
    }
  }

  payload.edges = payload.edges.filter(
    (e) => e.metadata?.direction !== 'reverse',
  );

  const normalizedEdges: GraphEdge[] = [];
  for (const edge of payload.edges) {
    const e = structuredClone(edge) as GraphEdge;
    let rel = e.relationship_type;
    if (!rel && e.Action) rel = actionToRel(e.Action);
    if (!rel) rel = actionToRel(String(e.metadata?.relationship_type ?? 'CONNECTED_TO'));
    e.relationship_type = rel;
    e.metadata = { ...(e.metadata ?? {}), ...(e.Action ? { action: e.Action } : {}) };
    normalizedEdges.push(e);
  }

  payload.edges = normalizedEdges;
  if (!payload.decision_id) {
    payload.decision_id = inferDecisionId(normalizedEdges, payload.updated_at);
  }

  stripOpenLifecycleOrphans(payload);
  stampCloseLifecycle(payload);
  scopeTradeLeafNodes(payload);
  stripCorrelatedOnlyMarketNodes(payload);
  stampAnchorMarketProperties(payload);

  payload.nodes = payload.nodes.map((n) => ({
    ...n,
    properties: n.properties ?? {},
    label: n.label ?? LABEL_MAP[n.node_type] ?? 'Entity',
  }));

  return payload;
}

export function escapeCypher(value: string): string {
  return value.replace(/\\/g, '\\\\').replace(/'/g, "\\'");
}

File: backend/app/lib/normalize.py
import copy
import re
from typing import Set

from app.lib.graph_topology import (
    agent_node_id,
    graph_id_for,
    parse_graph_id,
    primary_user_node_id,
    resolve_agent_context,
    user_slug_from_node_id,
)
from app.schemas.decision import DecisionEvent, GraphEdge, GraphNode

LEGACY_AGENT_ID = "publisher_agent"
AGENT_DISPLAY_ID = "Publisher Agent"

LABEL_MAP = {
    "user": "User",
    "protocol": "Protocol",
    "market": "Market",
    "correlated_market": "CorrelatedMarket",
    "trade": "Trade",
    "outcome": "Outcome",
    "feedback": "Feedback",
    "agent": "Agent",
}

LEGACY_CLOSE_TRADE_OF = {
    "TRD_003": "TRD_001",
    "TRD_005": "TRD_004",
    "TRD_007": "TRD_006",
    "TRD_010": "TRD_008",
    "TRD_012": "TRD_009",
    "TRD_014": "TRD_011",
    "TRD_017": "TRD_013",
}


def graph_name(graph_id: str) -> str:
    return re.sub(r"[^a-zA-Z0-9_]", "_", graph_id)


def escape_cypher(value: str) -> str:
    return value.replace("\\", "\\\\").replace("'", "\\'")


def scoped_leaf_id(leaf_id: str, trade_id: str) -> str:
    suffix = f"__{trade_id}"
    base = leaf_id
    while base.endswith(suffix):
        base = base[: -len(suffix)]
    return f"{base}{suffix}"


def _slug(value: str) -> str:
    return re.sub(r"^_+|_+$", "", re.sub(r"[^a-zA-Z0-9]+", "_", value)).lower()


def _action_to_rel(action: str) -> str:
    return action.strip().upper().replace(" ", "_")


def _infer_decision_id(edges: list[GraphEdge], updated_at: str) -> str:
    for edge in edges:
        if edge.Action and edge.target:
            return f"dec-{_slug(edge.target)}"
    return f"dec-{updated_at.replace(':', '').replace('.', '').replace('-', '')}"


def _remap_legacy_agent(payload: DecisionEvent, agent_id: str) -> None:
    node_ids = {n.node_id for n in payload.nodes}
    if LEGACY_AGENT_ID not in node_ids and agent_id not in node_ids:
        refs_legacy = any(
            e.source in (LEGACY_AGENT_ID, AGENT_DISPLAY_ID)
            or e.target in (LEGACY_AGENT_ID, AGENT_DISPLAY_ID)
            for e in payload.edges
        )
        if refs_legacy:
            payload.nodes.append(
                GraphNode(
                    node_id=agent_id,
                    node_type="agent",
                    properties={"display_name": "Publisher Agent", "role": "publisher"},
                    label="Agent",
                )
            )

    for node in payload.nodes:
        if node.node_id == LEGACY_AGENT_ID:
            node.node_id = agent_id

    for edge in payload.edges:
        if edge.source in (LEGACY_AGENT_ID, AGENT_DISPLAY_ID):
            edge.source = agent_id
        if edge.target in (LEGACY_AGENT_ID, AGENT_DISPLAY_ID):
            edge.target = agent_id
        if edge.targets:
            edge.targets = [
                agent_id if t in (LEGACY_AGENT_ID, AGENT_DISPLAY_ID) else t
                for t in edge.targets
            ]


def _remap_node_id(payload: DecisionEvent, from_id: str, to_id: str) -> None:
    node = next((n for n in payload.nodes if n.node_id == from_id), None)
    if not node or from_id == to_id:
        return
    node.node_id = to_id
    for edge in payload.edges:
        if edge.source == from_id:
            edge.source = to_id
        if edge.target == from_id:
            edge.target = to_id
        if edge.targets:
            edge.targets = [to_id if t == from_id else t for t in edge.targets]


def _collapse_agent_into_user(payload: DecisionEvent, user_node_id: str, agent_id: str) -> None:
    if user_node_id == agent_id:
        return
    for edge in payload.edges:
        if edge.source == agent_id:
            edge.source = user_node_id
        if edge.target == agent_id:
            edge.target = user_node_id
        if edge.targets:
            edge.targets = [user_node_id if t == agent_id else t for t in edge.targets]
    payload.edges = [e for e in payload.edges if e.relationship_type != "HAS_AGENT"]
    payload.nodes = [n for n in payload.nodes if n.node_id != agent_id and n.node_type != "agent"]


def _normalize_legacy_buy_sell_lifecycle(payload: DecisionEvent) -> None:
    for edge in payload.edges:
        if not edge.Action or not edge.target:
            continue
        match = re.match(r"^(Buy|Sell)\s+(YES|NO)$", edge.Action, re.I)
        if not match:
            continue
        verb = match.group(1).lower()
        side = match.group(2).upper()
        original_trade_id = edge.target
        is_close = verb == "sell"
        trade_id = LEGACY_CLOSE_TRADE_OF.get(original_trade_id, original_trade_id) if is_close else original_trade_id
        if trade_id != original_trade_id:
            _remap_node_id(payload, original_trade_id, trade_id)
        edge.Action = f"{'Close' if is_close else 'Open'} {side}"
        edge.relationship_type = _action_to_rel(edge.Action)
        edge.metadata = {
            **edge.metadata,
            "action": edge.Action,
            "lifecycle": "close" if is_close else "open",
            **({"position_trade_id": trade_id} if is_close else {}),
        }
        payload.operation = "revise" if is_close else "assert"
        payload.decision_id = f"dec-trd_{trade_id.replace('TRD_', '').lower()}-{'close' if is_close else 'open'}"


def _strip_open_lifecycle_orphans(payload: DecisionEvent) -> None:
    is_open = payload.operation == "assert" and any(
        e.metadata.get("lifecycle") == "open"
        or (e.Action and re.match(r"^Open\s", e.Action))
        for e in payload.edges
    )
    if not is_open:
        return
    leaf_ids = {n.node_id for n in payload.nodes if n.node_type == "feedback"}
    payload.edges = [e for e in payload.edges if not e.target or e.target not in leaf_ids]
    payload.nodes = [n for n in payload.nodes if n.node_type != "feedback"]


def _stamp_close_lifecycle(payload: DecisionEvent) -> None:
    is_close = payload.operation == "revise" or any(
        e.metadata.get("lifecycle") == "close"
        or (e.Action and re.match(r"^Close\s", e.Action))
        for e in payload.edges
    )
    if not is_close:
        return
    trade_node = next((n for n in payload.nodes if n.node_type == "trade"), None)
    if not trade_node:
        return
    for e in payload.edges:
        if e.Action and e.target == trade_node.node_id:
            e.metadata = {
                **e.metadata,
                "lifecycle": "close",
                "position_trade_id": trade_node.node_id,
            }


def _scope_trade_leaf_nodes(payload: DecisionEvent) -> None:
    trade_ids: Set[str] = set()
    for edge in payload.edges:
        if edge.Action and edge.target:
            trade_ids.add(edge.target)
    for trade_id in trade_ids:
        for edge in payload.edges:
            if edge.source != trade_id or not edge.target:
                continue
            target_node = next((n for n in payload.nodes if n.node_id == edge.target), None)
            if not target_node or target_node.node_type != "feedback":
                continue
            scoped_id = scoped_leaf_id(edge.target, trade_id)
            if edge.target != scoped_id:
                _remap_node_id(payload, edge.target, scoped_id)


def anchor_market_ids(payload: DecisionEvent) -> Set[str]:
    ids: Set[str] = set()
    for edge in payload.edges:
        if not edge.Action or not edge.target:
            continue
        trade = next(
            (n for n in payload.nodes if n.node_id == edge.target and n.node_type == "trade"),
            None,
        )
        if trade:
            ids.add(edge.source)
    return ids


def _correlated_target_ids(payload: DecisionEvent) -> Set[str]:
    ids: Set[str] = set()
    for edge in payload.edges:
        rel = (edge.relationship_type or str(edge.metadata.get("relationship_type", ""))).upper()
        rel = re.sub(r"[^A-Z0-9_]", "_", rel)
        if rel != "CORRELATED_MARKET":
            continue
        for tid in edge.targets or []:
            ids.add(tid)
    return ids


def _strip_correlated_only_market_nodes(payload: DecisionEvent) -> None:
    anchors = anchor_market_ids(payload)
    payload.nodes = [n for n in payload.nodes if n.node_type != "market" or n.node_id in anchors]


def _stamp_anchor_market_properties(payload: DecisionEvent) -> None:
    anchors = anchor_market_ids(payload)
    for node in payload.nodes:
        if node.node_type == "market" and node.node_id in anchors:
            node.properties = {**node.properties, "anchor": True}


def augment_correlated_peer_nodes(payload: DecisionEvent) -> DecisionEvent:
    existing = {n.node_id for n in payload.nodes}
    peers = _correlated_target_ids(payload)
    for node_id in peers:
        if node_id in existing:
            continue
        payload.nodes.append(
            GraphNode(
                node_id=node_id,
                node_type="correlated_market",
                properties={"correlated_peer": True},
                label="CorrelatedMarket",
            )
        )
        existing.add(node_id)
    return payload


def _ensure_has_agent_edge(payload: DecisionEvent, user_node_id: str, agent_id: str, role: str) -> None:
    node_ids = {n.node_id for n in payload.nodes}
    if user_node_id not in node_ids or agent_id not in node_ids:
        return
    has = any(
        e.source == user_node_id and e.target == agent_id and e.relationship_type == "HAS_AGENT"
        for e in payload.edges
    )
    if not has:
        payload.edges.insert(
            0,
            GraphEdge(
                source=user_node_id,
                target=agent_id,
                relationship_type="HAS_AGENT",
                metadata={"role": role},
            ),
        )


def normalize_decision(raw: DecisionEvent) -> DecisionEvent:
    payload = copy.deepcopy(raw)
    payload.schema_version = payload.schema_version or "1.0"
    payload.operation = payload.operation or "assert"
    _normalize_legacy_buy_sell_lifecycle(payload)

    ctx = resolve_agent_context(payload)
    if ctx:
        if not parse_graph_id(payload.graph_id):
            payload.graph_id = graph_id_for(ctx["userSlug"], ctx["role"])
        _remap_legacy_agent(payload, ctx["agentId"])
        if ctx["role"] == "seeker":
            _collapse_agent_into_user(payload, ctx["userNodeId"], ctx["agentId"])
        else:
            _ensure_has_agent_edge(payload, ctx["userNodeId"], ctx["agentId"], ctx["role"])
    else:
        user_node_id = primary_user_node_id(payload.nodes)
        if user_node_id:
            user_slug = user_slug_from_node_id(user_node_id)
            agent_id = agent_node_id(user_slug, "publisher")
            _remap_legacy_agent(payload, agent_id)
            _ensure_has_agent_edge(payload, user_node_id, agent_id, "publisher")

    payload.edges = [e for e in payload.edges if e.metadata.get("direction") != "reverse"]

    normalized_edges: list[GraphEdge] = []
    for edge in payload.edges:
        e = copy.deepcopy(edge)
        rel = e.relationship_type
        if not rel and e.Action:
            rel = _action_to_rel(e.Action)
        if not rel:
            rel = _action_to_rel(str(e.metadata.get("relationship_type", "CONNECTED_TO")))
        e.relationship_type = rel
        e.metadata = {**e.metadata, **({"action": e.Action} if e.Action else {})}
        normalized_edges.append(e)

    payload.edges = normalized_edges
    if not payload.decision_id:
        payload.decision_id = _infer_decision_id(normalized_edges, payload.updated_at)

    _strip_open_lifecycle_orphans(payload)
    _stamp_close_lifecycle(payload)
    _scope_trade_leaf_nodes(payload)
    _strip_correlated_only_market_nodes(payload)
    _stamp_anchor_market_properties(payload)

    payload.nodes = [
        GraphNode(
            node_id=n.node_id,
            node_type=n.node_type,
            properties=n.properties or {},
            label=n.label or LABEL_MAP.get(n.node_type, "Entity"),
        )
        for n in payload.nodes
    ]
    return payload
